src/java/org/apache/cassandra/repair/state/State.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.repair.state;
+
+import java.util.EnumMap;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.cassandra.utils.Clock;
+
+public interface State<T extends Enum<T>>
+{
+    UUID getId();
+
+    default boolean isComplete()
+    {
+        return getResult() != null;
+    }
+
+    T getStatus();
+
+    long getInitializedAtMillis();
+
+    long getInitializedAtNanos();
+
+    long getLastUpdatedAtMillis();
+
+    long getLastUpdatedAtNanos();
+
+    default long getDurationMillis()
+    {
+        long endNanos = getLastUpdatedAtNanos();
+        if (!isComplete())
+            endNanos = Clock.Global.nanoTime();
+        return TimeUnit.NANOSECONDS.toMillis(endNanos - getInitializedAtNanos());
+    }
+
+    float getProgress();
+
+    EnumMap<T, Long> getStateTimesMillis();
+
+    Result getResult();
+
+    default String getFailureCause()
+    {
+        Result r = getResult();
+        if (r == null || r.kind == Result.Kind.SUCCESS)
```

Comment:
why if the `state` is skipped? it will return skip-message instead of failure message

Reply:
I treat skip as a special case of failure, but from a message point of view they are the same thing as Result is just `kind` and `message`; so if you skip and ask for the failure cause, we give the skip message, this is intentional.